### ISSUE TITLE
Thoughts after reading LMDB docs again

### ISSUE THREAD

**Issue by andrewthad** [andrewthad/rotera#4]
I was going over the LMDB docs last night for inspiration. In LMDB, you have the remarkable property that separate processes (not just separate OS threads) can coordinate access to the database. On linux, it accomplishes this by keeping an interprocess posix mutex in the reader lock file. The readers all mmap the database file in read only mode (although they must all be able to write to the reader lock file). Since the readers and writer all share the same mmapped pages, you don't have to copy data to make it available. The operating system just pages in data as it's needed. Very cool. (You can also have the writer write directly to the mmapped memory, but this is dangerous.)

I was thinking about this mechanism and began to wonder if something like this would be appropriate for rotera. I eventually decided that I'm against attempting anything like this for the time being. Here are the concerns:

- Rotera doesn't currently support multi-process access. All concurrency is currently handled by GHC's STM. To support reads from multiple processes (where the readers open the database in read-only mode), we would need to use posix mutexes like LMDB does. We would also need to be able to signal these readers when new messages showed up in the queue so that we could have something like the `stream` command that the stream sockets API supports. (I really dislike the idea of using timers for this kind of thing, e.g. having the reader check back every N seconds.)
- How do we deal with an old reader? Currently, the only interface is stream sockets. If a reader is slowly reading some data that we want to overwrite because the queue has wrapped, we just `memcpy` that memory and keep `send`ing the copy. And GHC's STM makes it really easy to do this. But, if instead a reader process were sharing mmapped file, we would be in a rather different position. We couldn't just copy it. The reader process would be using a pointer into the mmapped region. How does LMDB handle this? In LMDB, everything is basically a big append-only B+ tree. By looking at the reader lock file, the writer is able to tell what [transaction the oldest reader is at](http://www.lmdb.tech/doc/group__readers.html). Any freed pages older than the oldest reader can be reclaimed. This works in LMDB because everything is segmented into pages (that's just how B+ trees work). It would not work for rotera because rotera uses a big rotating queue. To make this kind of thing possible in rotera, we would need to break the queue into segments, possibly 16MB each, and have the segments be ordered by an array somewhere else. While this would work, it's a totally different architecture than what we currently have. And users would only see any benefit it they put several readers on the same host as the writer.

In short, it would be rather complicated, and there is no immediate benefit. It's another project for another day. But if it were ever done, I believe that the stream-sockets interface to rotera might be able to be implemented as a reader process on top of this fancier architecture. In the case that multiple writers were needed, you'd use this process to coordinate the writers as well. In the case that only a single local writer was needed, we could take a page from LMDB's book and perform disk writes from the process doing the writing. (Well, if the stream-socket interface were used to coordinate writes, it would be the single local writer just doing this very thing.)

**Comment by andrewthad** [comment]
Today, I thought some more about other points in the design space.

There is a flag that [fallocate](http://man7.org/linux/man-pages/man2/fallocate.2.html) supports called `FALLOC_FL_COLLAPSE_RANGE`. One thing it can be used for is to effectively truncate at the beginning of a file instead of the end. Through experimentation, I have found that the positions of existing file handle are maintained as one would hope them to be. This could be used to build an event log where each of the readers just maintains an open handle and calls `read` to pull and logs while advancing the handle's position. However, there are some problems. Slow readers are difficult to deal with. Also, it's hard to figure out where the end of the queue is. If you just say "open a handle at the end of the file", I'm not sure if you might get placed in the middle of an event. Also, every read creates a new copy in memory since you aren't using a memory-mapped file anymore. If you went with this strategy, you would probably want to interleave sizes with logs. Anyway, it sounds cool in theory, but the practical problems outweigh the novelty.

I do feel more convinced that the best design for this is having a single file that's been segmented. Probably not the 16MB segments I previously suggested, but maybe more like 256MB or 1GB. With 256MB segments, if you capped the event log at 4TB, you could get away with a fixed-size array that holds metadata for the 16384 possible segments. I'm not sure exactly what this metadata would include though, but even if were 256 bytes, that would put you at just 4MB of fixed overhead. Not terrible, but not great either. What about the other metadata, the metadata for offsets. It would have to live in the same segment as the data it described (and events absolutely cannot cross segments). One kind of annoying thing we currently have to deal with is the problem of the offsets wrapping around (either the offsets or the events may currently wrap around, which guarantees that we are wasting space). But I believe there's a more elegant solution here that becomes available when we are working with segments. What if the events starting at the beginning of a segment, growing upward, and the offsets started at the end, growing downward? When they meet (which implies that almost no space is wasted), the segment is considered full. A full segment may no longer be written to until it is evicted. Eviction happens when there are no readers and when space is needed for a new segment. Oddly, a slow reader (or a hung reader) can cause a newer segment to be evicted before an older one, but this does not actually cause any strange behavior. Any scheme that deals with slow readers will have to make it possible for a reader to observe a big skip, and this is no exception. The important thing is that no new readers are allowed to start reading from segment being held up by the slow reader.

Segments mean that growing and shrinking are possible, something that `rotera` currently does not allow. Hole punching with `FALLOC_FL_PUNCH_HOLE` makes a shrink possible while plain old `fallocate` makes it possible to grow. Allocating blocks brings the possibility of out-of-disk-space errors, so I'm not sure if it is actually a good idea to let this happen on demand.

Anyway, still all a separate project for a separate day, but I wanted to get these notes down somewhere.
